browser/static/js/adjust.js

```diff
@@ -1,277 +1,376 @@
+    this.canvas.height = height;
+    this.canvas.width = width;
+    document.body.appendChild(this.canvas);
 
-function invert(img) {
-  let data = img.data;
-  for (var i = 0; i < data.length; i += 4) {
-    data[i]     = 255 - data[i];     // red
-    data[i + 1] = 255 - data[i + 1]; // green
-    data[i + 2] = 255 - data[i + 2]; // blue
-    }
-  return img;
-}
+    this.ctx = this.canvas.getContext("2d");
+    this.ctx.imageSmoothingEnabled = false;
+
+    // these will never change once initialized
+    this.height = height;
+    this.width = width;
+
+    this.rgb = rgb;
+
+    // TODO: also want invertMap for better adjustments
+    // when toggling between different channels
+    this.displayInvert = true;
 
-function preCompositeLabelMod(img, h1, h2) {
-  let ann = img.data;
-  // use label array to figure out which pixels to recolor
-  for (var j = 0; j < seg_array.length; j += 1){ //y
-    for (var i = 0; i < seg_array[j].length; i += 1){ //x
-      let jlen = seg_array[j].length;
-      let currentVal = Math.abs(seg_array[j][i]);
-      if (currentVal === h1 || currentVal === h2) {
-        recolorScaled(ann, i, j, jlen, r=255, g=-255, b=-255);
+    // brightness and contrast adjustment
+    // TODO: have this returned by a getter w/o a setter?
+    this.MIN_CONTRAST = -100;
+    this.MAX_CONTRAST = 700;
+
+    this.contrastMap = new Map();
+    this.brightnessMap = new Map();
+
+    for (let i = 0; i < channelMax; i++) {
+        this.brightnessMap.set(i, 0);
+        this.contrastMap.set(i, 0);
       }
+    this.brightness = this.brightnessMap.get(0);
+    this.contrast = this.contrastMap.get(0);
+
+    // raw and adjusted image storage
+    // cascasding image updates if raw or seg is reloaded
+    this.rawImage = new Image();
+    this.contrastedRaw = new Image();
+    this.preCompRaw = new Image();
+
+    this.segImage = new Image();
+    this.preCompSeg = new Image();
+
+    // adjusted raw + annotations
+    this.compositedImg = new Image();
+
+    // composite image + outlines, transparent highlight
+    this.postCompImg = new Image();
+
+    if (rgb) {
+      this.rawImage.onload = () => this.contrastRaw();
+      this.contrastedRaw.onload = () => this.rawAdjust(current_highlight, edit_mode, brush, mode);
+      this.segImage.onload = () => this.preCompAdjust(current_highlight, edit_mode, brush, mode);
+      this.preCompSeg.onload = () => this.segAdjust(current_highlight, edit_mode, brush, mode);
+    } else {
+      this.rawImage.onload = () => this.contrastRaw();
+      this.contrastedRaw.onload = () => this.preCompRawAdjust();
+      this.preCompRaw.onload = () => this.rawAdjust(current_highlight, edit_mode, brush, mode);
+      this.segImage.onload = () => this.preCompAdjust(current_highlight, edit_mode, brush, mode);
+      this.preCompSeg.onload = () => this.segAdjust(current_highlight, edit_mode, brush, mode);
+      this.compositedImg.onload = () => this.postCompAdjust(edit_mode, brush);
```

Comment:
`current_highlight`, `edit_mode`, `brush`, and `mode` are all undefined.

Reply:
Those variables are still globally scoped--I think we will need to iterate on how these pieces interact with each other as the code becomes more modular.

Reply:
Since they are all only used in the constructor, they could just be passed in.

Reply:
I gave this a try, but passing in the variables in the constructor doesn't result in the correct behavior.

Reply:
I think in the future, we may be able to store relevant settings in `adjuster` attributes, and have them update when the settings change. But I think that's out of scope for this PR and will likely depend on other parts of the code (especially `mode`) being refactored first so it doesn't get entirely out of hand.